walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,114 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.stream.Stream;
+
 
 public final class FindMeetingQuery {
+  public static final int MIN_INCREMENT = 10;
+
+  /** 
+  * Returns a Collection of {@code TimeRange} objects that span any available meeting times for all people in the MeetingRequest request.
+  * For example, if 4pm - 8pm is free for a meeting, 
+  * the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    // Mandatory events = events that must be attended by an attended in the Meeting Request
+    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
+    Collection<TimeRange> meetingTimes = findMeetingTimes(mtngDuration, mandatoryEvents);
+    return meetingTimes;
+  }
+
+  /** 
+  * Creates a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //check if any people in meeting request is present in this set
+      boolean mtngAttendeeInEvent = attendees.stream().anyMatch(attendee -> event.getAttendees().contains(attendee));
+
+      if (mtngAttendeeInEvent) {
+        // If there's already an event with this start time in the map, but the current end time is later, add the current instead.  
+        int eventStartTime = event.getWhen().start();
+        int eventEndTime = event.getWhen().end();
+        if (mandatoryEvents.containsKey(eventStartTime)) {
+          if (eventEndTime > mandatoryEvents.get(eventStartTime)) {
+            mandatoryEvents.put(eventStartTime, eventEndTime);
+          }
+        } else {
+          mandatoryEvents.put(eventStartTime, eventEndTime);
+        }
+      }
+    }
+    return mandatoryEvents;
+  }
+
+  /**
+  * Creates and returns an ArrayList of TimeRange objects that span available meeting times.
```

Comment:
```suggestion
  * Creates and returns an ArrayList of {@code TimeRange} objects that span available meeting times.
```

Reply:
Even better: Use `{@link TimeRange}` so that relevant tools will make this a link to the TimeRange class definition. Also, `{@link }`-ify `ArrayList`.